File: clojure/ch02/ex2.49.clj
(ns sicp.ch02 (:use clojure.test))

(defn make-vect [x y] [x y])
  
(defn xcor-vect [v] (first v))

(defn ycor-vect [v] (nth v 1))
  
(defn add-vect [v1 v2]
  (make-vect (+ (xcor-vect v1) (xcor-vect v2))
             (+ (ycor-vect v1) (ycor-vect v2))))

(defn sub-vect [v1 v2]
  (make-vect (- (xcor-vect v1) (xcor-vect v2))
             (- (ycor-vect v1) (ycor-vect v2))))

(defn scale-vect [factor v]
  (make-vect (* factor (xcor-vect v))
             (* factor (ycor-vect v))))

(defn make-segment [start-point end-point]  
  [start-point end-point])

(defn start-segment [s] (first s))

(defn end-segment [s] (nth s 1))

(defn make-frame [origin edge1 edge2]
  [origin edge1 edge2])

(defn origin-frame [frame]
  (nth frame 0))

(defn edge1-frame [frame]
  (nth frame 1))

(defn edge2-frame [frame]
  (nth frame 2))

; For the exercise, draw-line appends the line end points to a string
(def output (atom ""))
(defn draw-line [start end]
  (reset! output (format "%s[%s,%s]" @output start end)))
(defn reset-output [] (reset! output ""))

(defn frame-coord-map [frame]
  (fn [v]
    (add-vect
      (origin-frame frame)
      (add-vect (scale-vect (xcor-vect v)
                            (edge1-frame frame))
                (scale-vect (ycor-vect v)
                            (edge2-frame frame))))))
                            
(defn segments->painter [segment-list]
  (fn [frame]
    (loop [segments segment-list]
      (if (empty? segments) 
        '()
        (let [segment (first segments)]
          (draw-line
            ((frame-coord-map frame) (start-segment segment))
            ((frame-coord-map frame) (end-segment segment)))
          (recur (rest segments)))))))
                    
(def origin-1 (make-vect  0 0))
(def edge1-1  (make-vect  1 0)) ; edges are not relative to the origin values.
(def edge2-1  (make-vect  0 1))

(def origin-2 (make-vect -1 1))
(def edge1-2  (make-vect  2 1))
(def edge2-2  (make-vect -1 1))

(defn outline->painter [frame]
  (let [zero-zero-to-one-zero (make-segment (make-vect 0 0) (make-vect 1 0))
        one-zero-to-one-one   (make-segment (make-vect 1 0) (make-vect 1 1))
        one-one-to-zero-one   (make-segment (make-vect 1 1) (make-vect 0 1))
        zero-one-to-zero-zero (make-segment (make-vect 0 1) (make-vect 0 0))]
    ((segments->painter [zero-zero-to-one-zero one-zero-to-one-one 
                         one-one-to-zero-one zero-one-to-zero-zero]) frame)))
                            
(deftest test-outline->painter
  (outline->painter (make-frame origin-1 edge1-1 edge2-1))
  (is (= @output "[[0 0],[1 0]][[1 0],[1 1]][[1 1],[0 1]][[0 1],[0 0]]"))
  (reset-output)
  (outline->painter (make-frame origin-2 edge1-1 edge2-1))
  (is (= @output "[[-1 1],[0 1]][[0 1],[0 2]][[0 2],[-1 2]][[-1 2],[-1 1]]"))
  (reset-output)
  (outline->painter (make-frame origin-1 edge1-2 edge2-2))
  (is (= @output "[[0 0],[2 1]][[2 1],[1 2]][[1 2],[-1 1]][[-1 1],[0 0]]"))
  (reset-output)
  (outline->painter (make-frame origin-2 edge1-2 edge2-2))
  (is (= @output "[[-1 1],[1 2]][[1 2],[0 3]][[0 3],[-2 2]][[-2 2],[-1 1]]"))
  (reset-output))

(defn x->painter [frame]
  (let [zero-zero-to-one-one (make-segment (make-vect 0 0) (make-vect 1 1))
        one-zero-to-zero-one   (make-segment (make-vect 1 0) (make-vect 0 1))]
    ((segments->painter [zero-zero-to-one-one one-zero-to-zero-one]) frame)))

(deftest test-x->painter
  (x->painter (make-frame origin-1 edge1-1 edge2-1))
  (is (= @output "[[0 0],[1 1]][[1 0],[0 1]]"))
  (reset-output)
  (x->painter (make-frame origin-2 edge1-1 edge2-1))
  (is (= @output "[[-1 1],[0 2]][[0 1],[-1 2]]"))
  (reset-output)
  (x->painter (make-frame origin-1 edge1-2 edge2-2))
  (is (= @output "[[0 0],[1 2]][[2 1],[-1 1]]"))
  (reset-output)
  (x->painter (make-frame origin-2 edge1-2 edge2-2))
  (is (= @output "[[-1 1],[0 3]][[1 2],[-2 2]]"))
  (reset-output))

(defn diamond->painter [frame]
  (let [one   (make-segment (make-vect 0.5 0) (make-vect 1 0.5))
        two   (make-segment (make-vect 1 0.5) (make-vect 0.5 1))
        three (make-segment (make-vect 0.5 1) (make-vect 0 0.5))
        four  (make-segment (make-vect 0 0.5) (make-vect 0.5 0))]
    ((segments->painter [one two three four]) frame)))

(deftest test-diamond->painter
  (diamond->painter (make-frame origin-1 edge1-1 edge2-1))
  (is (= @output "[[0.5 0.0],[1.0 0.5]][[1.0 0.5],[0.5 1.0]][[0.5 1.0],[0.0 0.5]][[0.0 0.5],[0.5 0.0]]"))
  (reset-output)
  (diamond->painter (make-frame origin-2 edge1-1 edge2-1))
  (is (= @output "[[-0.5 1.0],[0.0 1.5]][[0.0 1.5],[-0.5 2.0]][[-0.5 2.0],[-1.0 1.5]][[-1.0 1.5],[-0.5 1.0]]"))
  (reset-output)
  (diamond->painter (make-frame origin-1 edge1-2 edge2-2))
  (is (= @output "[[1.0 0.5],[1.5 1.5]][[1.5 1.5],[0.0 1.5]][[0.0 1.5],[-0.5 0.5]][[-0.5 0.5],[1.0 0.5]]"))
  (reset-output)
  (diamond->painter (make-frame origin-2 edge1-2 edge2-2))
  (is (= @output "[[0.0 1.5],[0.5 2.5]][[0.5 2.5],[-1.0 2.5]][[-1.0 2.5],[-1.5 1.5]][[-1.5 1.5],[0.0 1.5]]"))
  (reset-output))

(run-tests)

; skipped the wave painter

File: scala/ch02/ex2.49.scala
type Vect = Tuple2[Double,Double]

def makeVect (x:Double, y:Double) = (x, y)
  
def xcorVect (v:Vect) = v._1

def ycorVect (v:Vect) = v._2
  
def addVect (v1:Vect, v2:Vect) =
  makeVect (xcorVect(v1) + xcorVect(v2),
            ycorVect(v1) + ycorVect(v2))

def subVect (v1:Vect, v2:Vect) =
  makeVect (xcorVect(v1) - xcorVect(v2),
            ycorVect(v1) - ycorVect(v2))

def scaleVect (factor:Double, v:Vect) =
  makeVect (factor * xcorVect(v),
            factor * ycorVect(v))

// If startPoint and endPoint are each (x,y) pairs, then they are already vectors
// as defined by makeVect above. We make a list of the points.

type Segment = Tuple2[Vect,Vect]

def makeSegment (startPoint: Vect, endPoint: Vect) = (startPoint, endPoint)

def startSegment (s: Segment) = s._1

def endSegment (s: Segment) = s._2
  
type Frame = Tuple3[Vect,Vect,Vect]

def makeFrame (origin:Vect, edge1:Vect, edge2:Vect) = 
  (origin, edge1, edge2)

def originFrame (frame:Frame) = frame._1

def edge1Frame (frame:Frame) = frame._2

def edge2Frame (frame:Frame) = frame._3

// For the exercise, drawLine appens the pair of the line ends to a string.
var output = ""
def drawLine (start: Vect, end: Vect) = output += "(" + start + "," + end + ")"
def resetOutput = output = ""

def frameCoordMap (frame: Frame) = (v: Vect) =>
  addVect(
    originFrame(frame),
    addVect (scaleVect (xcorVect(v), edge1Frame(frame)),
             scaleVect (ycorVect(v), edge2Frame(frame))))
                            
def segmentsPainter (segments: List[Segment]) = (frame: Frame) =>
  segments foreach { segment => 
    drawLine (frameCoordMap(frame) (startSegment(segment)),
              frameCoordMap(frame) (endSegment(segment)))
  }
          
def outlinePainter (frame: Frame) = {
  val zeroZeroToOneZero = makeSegment (makeVect(0, 0), makeVect(1, 0))
  val oneZeroToOneOne   = makeSegment (makeVect(1, 0), makeVect(1, 1))
  val oneOneToZeroOne   = makeSegment (makeVect(1, 1), makeVect(0, 1))
  val zeroOneToZeroZero = makeSegment (makeVect(0, 1), makeVect(0, 0))
  segmentsPainter (zeroZeroToOneZero :: oneZeroToOneOne :: 
                   oneOneToZeroOne :: zeroOneToZeroZero :: Nil) (frame)
}
                            
val origin1 = makeVect(0, 0)
val edge11  = makeVect(1, 0) // edges are not relative to the origin values.
val edge21  = makeVect(0, 1)

val origin2 = makeVect(-1, 1)
val edge12  = makeVect( 2, 1)
val edge22  = makeVect(-1, 1)

import org.scalatest._ 
import org.scalatest.matchers._

object outlinePainterSpec extends Spec with ShouldMatchers {
  describe ("outlinePainter") {
    it ("should trace the corners of the frame") {
      outlinePainter (makeFrame (origin1, edge11, edge21))
      output should equal ("((0.0,0.0),(1.0,0.0))((1.0,0.0),(1.0,1.0))((1.0,1.0),(0.0,1.0))((0.0,1.0),(0.0,0.0))")
      resetOutput
      outlinePainter (makeFrame (origin2, edge11, edge21))
      output should equal ("((-1.0,1.0),(0.0,1.0))((0.0,1.0),(0.0,2.0))((0.0,2.0),(-1.0,2.0))((-1.0,2.0),(-1.0,1.0))")
      resetOutput
      outlinePainter (makeFrame (origin1, edge12, edge22))
      output should equal ("((0.0,0.0),(2.0,1.0))((2.0,1.0),(1.0,2.0))((1.0,2.0),(-1.0,1.0))((-1.0,1.0),(0.0,0.0))")
      resetOutput
      outlinePainter (makeFrame (origin2, edge12, edge22))
      output should equal ("((-1.0,1.0),(1.0,2.0))((1.0,2.0),(0.0,3.0))((0.0,3.0),(-2.0,2.0))((-2.0,2.0),(-1.0,1.0))")
      resetOutput
    }
  }
}
outlinePainterSpec execute

def xPainter (frame: Frame) = {
  val zeroZeroToOneOne = makeSegment (makeVect(0, 0), makeVect(1, 1))
  val oneZeroToZeroOne = makeSegment (makeVect(1, 0), makeVect(0, 1))
  segmentsPainter (zeroZeroToOneOne :: oneZeroToZeroOne :: Nil) (frame)
}
                            
object xPainterSpec extends Spec with ShouldMatchers {
  describe ("xPainter") {
    it ("should trace an 'x' between the corners of the frame") {
      xPainter (makeFrame (origin1, edge11, edge21))
      output should equal ("((0.0,0.0),(1.0,1.0))((1.0,0.0),(0.0,1.0))")
      resetOutput
      xPainter (makeFrame (origin2, edge11, edge21))
      output should equal ("((-1.0,1.0),(0.0,2.0))((0.0,1.0),(-1.0,2.0))")
      resetOutput
      xPainter (makeFrame (origin1, edge12, edge22))
      output should equal ("((0.0,0.0),(1.0,2.0))((2.0,1.0),(-1.0,1.0))")
      resetOutput
      xPainter (makeFrame (origin2, edge12, edge22))
      output should equal ("((-1.0,1.0),(0.0,3.0))((1.0,2.0),(-2.0,2.0))")
      resetOutput
    }
  }
}
xPainterSpec execute

def diamondPainter (frame: Frame) = {
  val one   = makeSegment (makeVect(0.5, 0), makeVect(1, 0.5))
  val two   = makeSegment (makeVect(1, 0.5), makeVect(0.5, 1))
  val three = makeSegment (makeVect(0.5, 1), makeVect(0, 0.5))
  val four  = makeSegment (makeVect(0, 0.5), makeVect(0.5, 0))
  segmentsPainter (one :: two :: three :: four :: Nil) (frame)
}
                            
object diamondPainterSpec extends Spec with ShouldMatchers {
  describe ("diamondPainter") {
    it ("should trace a diamond in the frame") {
      diamondPainter (makeFrame (origin1, edge11, edge21))
      output should equal ("((0.5,0.0),(1.0,0.5))((1.0,0.5),(0.5,1.0))((0.5,1.0),(0.0,0.5))((0.0,0.5),(0.5,0.0))")
      resetOutput
      diamondPainter (makeFrame (origin2, edge11, edge21))
      output should equal ("((-0.5,1.0),(0.0,1.5))((0.0,1.5),(-0.5,2.0))((-0.5,2.0),(-1.0,1.5))((-1.0,1.5),(-0.5,1.0))")
      resetOutput
      diamondPainter (makeFrame (origin1, edge12, edge22))
      output should equal ("((1.0,0.5),(1.5,1.5))((1.5,1.5),(0.0,1.5))((0.0,1.5),(-0.5,0.5))((-0.5,0.5),(1.0,0.5))")
      resetOutput
      diamondPainter (makeFrame (origin2, edge12, edge22))
      output should equal ("((0.0,1.5),(0.5,2.5))((0.5,2.5),(-1.0,2.5))((-1.0,2.5),(-1.5,1.5))((-1.5,1.5),(0.0,1.5))")
      resetOutput
    }
  }
}
diamondPainterSpec execute

// skipped d) the wave painter.
